Crash with multiple video tags that contains webm source

jyavenard opened the pull request:
#### 75467caf79bb0cbbc7441ecbe4a239829779eac3
<pre>
Crash with multiple video tags that contains webm source
<a href="https://bugs.webkit.org/show_bug.cgi?id=276717">https://bugs.webkit.org/show_bug.cgi?id=276717</a>
<a href="https://rdar.apple.com/132405520">rdar://132405520</a>

Reviewed by Youenn Fablet.

The WebM MediaPlayer ignored the preload attribute and would read the webm video
in its entirety all the time. This could lead to great memory usage and eventually the
page being jetsammed.
We now stops loading the video if preload is set to &lt;= metadata and will continue
from where we stopped when play() is called.
The `requestMediaDataWhenReady` callback was set whenever we demuxed a
single sample or when a new audio or video renderer was created, flushed etc.
It only needed to be set once.
The dimensions of the video were determined after enqueuing the first video sample.
We now read it from the container instead as it allows to reduce how much of the
media we have to download to obtain the information and move the readyState to
HAVE_METADATA.

Added test.

* LayoutTests/media/media-vp8-webm-withp-reload-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-with-preload.html: Added.
* LayoutTests/platform/glib/TestExpectations: see webkit.org/b/277327
* LayoutTests/platform/mac-wk1/TestExpectations: webm not supported in wk1.
* Source/WebCore/platform/graphics/WebMResourceClient.cpp:
(WebCore::WebMResourceClient::responseReceived):
* Source/WebCore/platform/graphics/WebMResourceClient.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::setPreload):
(WebCore::MediaPlayerPrivateWebM::doPreload):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::createResourceClient):
(WebCore::MediaPlayerPrivateWebM::dataLengthReceived):
(WebCore::MediaPlayerPrivateWebM::dataReceived):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::prepareToPlay):
(WebCore::MediaPlayerPrivateWebM::setNaturalSize):
(WebCore::MediaPlayerPrivateWebM::setNetworkState):
(WebCore::MediaPlayerPrivateWebM::maybeFinishLoading):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::didProvideMediaDataForTrackId):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::configureVideoRenderer):
(WebCore::MediaPlayerPrivateWebM::setVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.h: We were relying on the first video frame
being enqueued to determine the dimensions of the video. We&apos;ll now instead rely on the container&apos;s track data.
So we make public the width and height function.

Canonical link: <a href="https://commits.webkit.org/281554@main">https://commits.webkit.org/281554@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0de67b14ea26bfe31283e8be8f9b4126394b91d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60311 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64231 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10843 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62441 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11076 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48832 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7548 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62342 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36970 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29678 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33669 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/9479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9760 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9767 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65963 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4245 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56193 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56360 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13382 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3536 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/36298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -198,33 +198,68 @@ static bool isCopyDisplayedPixelBufferAvailable()
+        request.addHTTPHeaderField(HTTPHeaderName::Range, makeString("bytes="_s, m_contentReceived, '-', m_contentLength));
+    }
 
     auto loader = player->createResourceLoader();
```
It would be good to do a refactoring to return a Ref<>

jyavenard replied:
I'll take this in a follow-up

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -198,33 +198,68 @@ static bool isCopyDisplayedPixelBufferAvailable()
+
+    m_assetURL = URL({ }, url);
+
+    setPreload(m_preload);
```
We trigger again preload on change of URL, but we do a !m_resourceClient which might be already created?
Do we have tests that cover this code path?

jyavenard replied:
Changing the URL in an existing MediaPlayerPrivateWebM is not going to work, even prior any of those changes.

The SourceBufferParser used to demux the webm is created at the player construction, it can't handle a change of content mid-way.

If this is to be supported, it's currently broken

jyavenard replied:
A new `MediaPlayer` is created whenever the `src` attribute is modified.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -198,33 +198,68 @@ static bool isCopyDisplayedPixelBufferAvailable()
+        return false;
+
     ResourceRequest request(url);
     request.setAllowCookies(true);
```
This is existing code but it feels a bit weird that this is handled there and not for container agnostic code.

eric-carlson replied:
+1

jyavenard replied:
I'm not sure I understand the comment here. 
m_player is a WeakPtr, and we need to access it to create the ResourceLoader.

jyavenard replied:
the weakptr to the player is owned by each MediaPlayerPrivate ; not checking there and have it be owned at the higher level would be a big change, including having to typically downcast in some circumstances.

youennf replied:
It is the call to setAllowCookies that feels a bit out of place, this could be best addressed in a follow-up.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -198,33 +198,68 @@ static bool isCopyDisplayedPixelBufferAvailable()
 }
 
-void MediaPlayerPrivateWebM::load(const String& url)
+void MediaPlayerPrivateWebM::setPreload(MediaPlayer::Preload preload)
```
This method is doing much more than setting preload, it is actually preloading if needed.
Maybe the name should be clarified (and yes the setPreload overriding method would call this new method).
Something like `doPreload()`

jyavenard replied:
setPreload is the name of the virtual method, I copied the logic and the code from the MediaPlayerPrivateAVFoundation

eric-carlson replied:
I wouldn’t assume everything `MediaPlayerPrivateAVFoundation` does is correct

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -213,18 +224,51 @@ static bool isCopyDisplayedPixelBufferAvailable()
+    doPreload();
+}
+
+bool MediaPlayerPrivateWebM::createResourceClient(const URL& url)
```
I would just not pass url here given it can only be m_assetURL.

jyavenard committed:
Crash with multiple video tags that contains webm source
https://bugs.webkit.org/show_bug.cgi?id=276717
rdar://132405520

Reviewed by Youenn Fablet.

The WebM MediaPlayer ignored the preload attribute and would read the webm video
in its entirety all the time. This could lead to great memory usage and eventually the
page being jetsammed.
We now stops loading the video if preload is set to <= metadata and will continue
from where we stopped when play() is called.
The `requestMediaDataWhenReady` callback was set whenever we demuxed a
single sample or when a new audio or video renderer was created, flushed etc.
It only needed to be set once.
The dimensions of the video were determined after enqueuing the first video sample.
We now read it from the container instead as it allows to reduce how much of the
media we have to download to obtain the information and move the readyState to
HAVE_METADATA.

Added test.

* LayoutTests/media/media-vp8-webm-withp-reload-expected.txt: Added.
* LayoutTests/media/media-vp8-webm-with-preload.html: Added.
* LayoutTests/platform/glib/TestExpectations: see webkit.org/b/277327
* LayoutTests/platform/mac-wk1/TestExpectations: webm not supported in wk1.
* Source/WebCore/platform/graphics/WebMResourceClient.cpp:
(WebCore::WebMResourceClient::responseReceived):
* Source/WebCore/platform/graphics/WebMResourceClient.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::setPreload):
(WebCore::MediaPlayerPrivateWebM::doPreload):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::createResourceClient):
(WebCore::MediaPlayerPrivateWebM::dataLengthReceived):
(WebCore::MediaPlayerPrivateWebM::dataReceived):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::prepareToPlay):
(WebCore::MediaPlayerPrivateWebM::setNaturalSize):
(WebCore::MediaPlayerPrivateWebM::setNetworkState):
(WebCore::MediaPlayerPrivateWebM::maybeFinishLoading):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::didProvideMediaDataForTrackId):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::configureVideoRenderer):
(WebCore::MediaPlayerPrivateWebM::setVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.h: We were relying on the first video frame
being enqueued to determine the dimensions of the video. We'll now instead rely on the container's track data.
So we make public the width and height function.

Canonical link: https://commits.webkit.org/281554@main
